DependencyInjection/Compiler/EntityListenerPass.php

```diff
@@ -87,40 +88,57 @@ public function process(ContainerBuilder $container)
         }
     }
 
-    private function attachToListener(ContainerBuilder $container, $name, $id, array $attributes)
+    private function attachToListener(ContainerBuilder $container, $name, string $class, array $attributes)
     {
         $listenerId = sprintf('doctrine.orm.%s_listeners.attach_entity_listeners', $name);
 
         if (! $container->has($listenerId)) {
             return;
         }
 
-        $serviceDef = $container->getDefinition($id);
-
         $args = [
             $attributes['entity'],
-            $serviceDef->getClass(),
+            $class,
             $attributes['event'],
         ];
 
         if (isset($attributes['method'])) {
             $args[] = $attributes['method'];
-        } elseif (! method_exists($serviceDef->getClass(), $attributes['event']) && method_exists($serviceDef->getClass(), '__invoke')) {
+        } elseif (! method_exists($class, $attributes['event']) && method_exists($class, '__invoke')) {
             $args[] = '__invoke';
         }
 
         $container->findDefinition($listenerId)->addMethodCall('addEntityListener', $args);
     }
 
-    private function getResolverClass(Definition $resolver, ContainerBuilder $container) : string
+    private function getResolverClass(Definition $resolver, ContainerBuilder $container, string $id) : string
     {
-        $resolverClass = $resolver->getClass();
+        $resolverClass = $this->getConcreteDefinitionClass($resolver, $container, $id);
 
         if (substr($resolverClass, 0, 1) === '%') {
             // resolve container parameter first
-            $resolverClass = $container->getParameterBag()->resolveValue($resolver->getClass());
+            $resolverClass = $container->getParameterBag()->resolveValue($resolverClass);
         }
 
         return $resolverClass;
     }
+
+    private function getConcreteDefinitionClass(Definition $definition, ContainerBuilder $container, string $id) : string
+    {
+        $class = $definition->getClass();
+        if (is_string($class)) {
```

Comment:
we don't usually check with `is_string` in similar situations, it should be just `if ($class) {` to me

Reply:
@fancyweb Do you have this change request on your todo list?

Reply:
Just done it.